### PR TITLE
Prefilling Room Database with JSON Data in Android

### LINKED ISSUE #6926: [Android] Prefilling Room Database with JSON Data in Android
#### Proposed article introduction
Sometimes, you might want your app to start with a database that is already loaded with a specific set of data. This is called pre-populating a database. In Room 2.2.0 and higher, you can use API methods to pre-populate a Room database during initialization with content from a prepackaged database (.db)  file in the device's file system.

From the Android documentation:
- To pre-populate a Room database from a prepackaged database file that is located anywhere in your app's assets/ directory, you can call the createFromAsset() method.
- To pre-populate a Room database from a prepackaged database file that is located anywhere in the device's file system except your app's assets/ directory, call the createFromFile() method.

From the docs, there is no indication as to how you can prefill your DB with JSON data. In this tutorial, we are going to prefill the database that we are going to create with some JSON data that I have prepared, from your end you could be filling it with some JSON data that you have downloaded from your backend.

#### Key takeaways 
- A recap of setting up the Room database.
- Understanding what pre-populating Room database with initial data is.
- When is it necessary to pre-populate a Room DB?
- What is a Room DB callback and how do I add a callback to the database?
- Implementing the pre-filling of Room DB.

#### Article quality
Most of the blog sites and tutorials online focus on pre-populating room database with `.db` data which matches the schema of the database with `createFromAssets` and  `createFromFile` methods. In this tutorial I will make it unique by coming up with an easy and straightforward method of pre-populating Room DB with JSON data.

#### References
- N/A

### PR DESCRIPTION
**Pull request by benta-odek**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #6926

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by louisefindlay23** [comment]
Plagiarism check complete. Please rename your branch and folder name to use lowercase.

**Comment by benta-odek** [comment]
Hello @louisefindlay23 have fixed the issue

**Comment by hectorkambow** [comment]
cc @BrianaNzivu 

any update on this review? Lets be sure to glance at the content before we dive into our review to see if any glaring issue exist.

**Comment by BrianaNzivu** [comment]
> cc @BrianaNzivu
> 
> any update on this review? Lets be sure to glance at the content before we dive into our review to see if any glaring issue exist.

Working on it.

File: content/articles/prefilling-room-database-with-json-data-in-android/index.md [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+- [Displaying Data](#step-9---displaying-data)
+- [When is it Necessary to Pre-populate a Room Database](#when-is-it-necessary-to-pre-populate-a-room-database)
+- [Conclusion](#conclusion)
+
+### Prerequisites
+To read along with this tutorial you should:-
+- Have a fundamental grasp of Android programming.
+- Have a thorough knowledge of the Kotlin programming language.
+- ViewModels concept as utilized in Android.
+
```
Comment (BrianaNzivu) [inline content/articles/prefilling-room-database-with-json-data-in-android/index.md:41]: I think you also need an IDE.

File: content/articles/prefilling-room-database-with-json-data-in-android/index.md [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+### Goals
+This tutorial aims to explain what is pre-populating room database with initial data as used in android. what is a room Callback is and how is it added to the database? How to implement prefilling room database with data and when it is necessary to pre-populate a room database.
```
Comment (BrianaNzivu) [inline content/articles/prefilling-room-database-with-json-data-in-android/index.md:43]: Have this in point form

**Review by BrianaNzivu — CHANGES_REQUESTED** [review]
Hey @benta-odek benta-odek👋 .

My name is Briana and I will be reviewing your article. I promise to make this review as fast and as swift as possible, with your help of course 😄 .

Good job on the article. It is a very interesting read.

Just a few changes I need you to make as we continue with the preliminary revisions:
- I want you to read through your article using [Grammarly](https://app.grammarly.com/) and correct any grammar mistakes.
- Please refer to the comments made.

In case of any question or help needed please do not hesitate to ping me here or on Slack.
Also, notify me when you are done.

Regards.

**Comment by benta-odek** [comment]
Hello @BrianaNzivu have made the requested changes

**Review by BrianaNzivu — APPROVED** [review]
Hey @benta-odek 👋 .

Very good job on the article. It is a very interesting read 👏 .

Just a few things I'd like to point out:
-  It is always advisable to submit an article that is well polished. By this, I mean having an article that does not have too many grammatical corrections and functional code. I suggest that you check out our [resources](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) page which has links to various resourceful tools for writing. 
- Lastly please go through ALL the corrections I have made so that you can not repeat the same mistakes.

That's it! :smile: 

A final review will be done now by a content moderator.
Otherwise, congratulations on publishing your article🍾.

In case of any question or clarification (currently or in the future), do not hesitate to reach out via [Slack](https://app.slack.com/client/T5X14SH7G/D01A6MT673R) or my personal email (briananzivu@gmail.com). 
I am **ALWAYS** here to help.

It has been a pleasure reading and reviewing this article.

Looking forward to seeing you contribute more to EngEd. 🚀

Kind Regards.

**Pull request merged**